data/src/history.rs

```diff
@@ -171,6 +188,144 @@ impl History {
         }
     }
 
+    fn add_chathistory_message(
+        &mut self,
+        message: Message,
+        subcommand: ChatHistorySubcommand,
+        message_reference: MessageReference,
+    ) {
+        match self {
```

Comment:
I haven't tested it, but I think we can simplify this and expose a single `add_message` API with the following. This is also non-dependent on `ChatHistorySubcommand`, not sure why that's needed here for splicing the message in.

We basically splice messages in `server_time` order, allowing us to `binary_search` on that time which should be pretty fast.

Binary search returns a `Result<usize, usize>` where the Ok value means an identical `Ordering::Equal` match was found, so we can swap out that value with the incoming value. 

This is 100% safe when we use the `Lookup::MessageId` since we return `Equal` if the id's match, otherwise the binary search continues based on server time.

For `Lookup::Timestamp` this means the message has an identical timestamp. Since `servertime` extension mandates milisecond accuracy, and if we don't have servertime, we use `Utc::now` which is nanosecond accuracy, I think this is a strong match. I doubt we will have too many false positives. 

The Err value means no match was found, but the position returns is the index where the element can be safely inserted to preserve ordering (server_time), so we can just safely insert it.

```rs
    fn add_message(&mut self, message: Message, reference: Option<MessageReference>) {
        if let History::Partial {
            unread_message_count,
            ..
        } = self
        {
            if message.triggers_unread() {
                *unread_message_count += 1;
            }
        }

        match self {
            History::Partial {
                messages,
                last_received_at,
                ..
            }
            | History::Full {
                messages,
                last_received_at,
                ..
            } => {
                enum Lookup {
                    Timestamp(DateTime<Utc>),
                    MessageId(String),
                }

                let lookup = match reference {
                    Some(MessageReference::Timestamp(timestamp, _)) => Lookup::Timestamp(timestamp),
                    Some(MessageReference::MessageId(id)) => Lookup::MessageId(id),
                    Some(MessageReference::None) | None => {
                        if let Some(id) = message.id.clone() {
                            Lookup::MessageId(id)
                        } else {
                            Lookup::Timestamp(message.server_time)
                        }
                    }
                };

                let search = match lookup {
                    Lookup::Timestamp(timestamp) => {
                        messages.binary_search_by(|stored| stored.server_time.cmp(&timestamp))
                    }
                    Lookup::MessageId(id) => messages.binary_search_by(|stored| {
                        if stored.id.as_ref() == Some(&id) {
                            Ordering::Equal
                        } else {
                            stored.server_time.cmp(&message.server_time)
                        }
                    }),
                };

                match search {
                    // Identical value found, replace it
                    //
                    // Message Id is 100% match
                    // Timestamp (server time) is milisecond accuracy so
                    // a very strong match
                    Ok(pos) => {
                        messages.splice(pos..pos + 1, Some(message));
                    }
                    // No match found, insert it at pos (server time sorted order)
                    Err(pos) => messages.insert(pos, message),
                }

                *last_received_at = Some(Instant::now());
            }
        }
    }
```

Reply:
I had considered something like this (was thinking of using a `BTreeMap`), but I was worried about the overhead it would put on the non-chathistory variant of `add_message`.

I'll put this in and test it.

Reply:
Binary search on 10k messages will be pretty quick, that's like 13-14 iterations before you've exhausted the search. 

However, I think I need to make it more robust. I have some ideas and will dig in more tomorrow.

Reply:
I believe we'd need to sort the messages by `server_time` or `id` for this to work.  Sorting by `server_time` should be okay, that's more or less how we expect messages to be already, but we have no guarantee that sorting by `id` will match the ordering given by `server_time`.

Reply:
Correct, the binary search will insert things in server time sorted order.

I plan to do 2 binary search lookups +- N duration from current message server time.

I can then iterator through those messages and check for an id match. If an id match occurs, replace. If no id match, I can insert after the last position where server time <= current message.

This allows us to handle checking messages with duplicate server times, which the binary search function on its own doesn't handle.

We can also check user + content as well if no id exists for duplicate.

Reply:
Another consideration the specification brings up is that messages do not have to be ordered by server time in the server message log.  This is a reason supporting their suggestion to use a "fuzz interval" when requesting new messages via timestamp.  This means client-side message ordering *might* not match server side.  I don't know that we want to do anything to accommodate that (beyond adding the fuzz interval to our `Timestamp` requests, which we currently do).

Reply:
Content is unfortunately tricky to check for matching for event messages, in my experience, but works well otherwise in my experience.

Reply:
If we just blindly append like previously, then we're back to doing linear search which is much worse performance wise than binary search. Plus we'd have to sort at presentation time if we wanted to enforce that strictly. 

I think the +- binary search to get a surrounding slice and then linear search through that slice approach I mentioned will work really well and give us both performance and robustness.

Reply:
Oh, absolutely, I didn't mean to suggest otherwise.  I was just accepting worse chathistory performance in order to avoid messing with the non-chathistory `add_message`.

Reply:
For sure. I'm thinking this add_message can be completely unaware of chat history since we already will have the id and server time in the incoming message, so we can just treat all incoming messages the same for how we insert them.

Reply:
Sounds good to me.  I will leave it to you :)

Reply:
Looks good to me!  I agree that checking all messages for a clash is not necessary (and potentially not desirable, since a non-duplicate clash means something odd has happened;  let's cross that bridge when we come to it).

I incorporated it further into the existing PR and made a couple small tweaks:
1.  I think we don't want to mark a message as incrementing the unread count when it is truly a duplicate of an existing message.  Timestamp-based requests are fuzzed, so we will frequently get a message back that the user has already seen.  So I added a bit of logic to not increment the unread message count in that scenario.
2.  `JOIN`/`PART`/`QUIT` messages frequently, in my experience, do not match well on text due to how the user attributes resolve.  E.g. when a new user joins a channel it's not uncommon for their user attributes to not resolve at all if they weren't in another channel with the Halloy user, but if they're still in the channel with the Halloy user when the `chathistory` batch is received then their user attributes will resolve when the `chathistory` copy of the join is received.  So I relaxed the match requirements for those messages to just match on the target, which contains the nick of the user involved in the message.

Reply:
Sounds good to me!